### PR TITLE
overlord: frontend for --quota-group support (2/2)

### PR DESCRIPTION
**Pull request by Meulengracht**
The second PR for the '--quota-group' option on snap install. This effectively invalidates the original PR https://github.com/snapcore/snapd/pull/11940 and is a followup to https://github.com/snapcore/snapd/pull/12066 which is also contained here.

This PR contains the frontend changes required to support the option, including a spread test.

### CODE REVIEW

File: daemon/api_sideload_n_try.go [hunk 1]
```diff
@@ -154,6 +154,13 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string, user *au
 		return trySnap(c.d.overlord.State(), form.Values["snap-path"][0], flags)
 	}
 
+	if len(form.Values["quota-group"]) > 0 {
+		if len(form.Values["quota-group"]) != 1 {
+			return BadRequest("too many names provided for 'quota-group' option")
+		}
+		flags.QuotaGroupName = form.Values["quota-group"][0]
+	}
```
Comment (pedronis) [inline daemon/api_sideload_n_try.go:162]: this needs a unit test?

File: daemon/api_snaps.go [hunk 2]
```diff
@@ -203,6 +203,7 @@ type snapInstruction struct {
 	Snaps                  []string               `json:"snaps"`
 	Users                  []string               `json:"users"`
 	ValidationSets         []string               `json:"validation-sets"`
+	QuotaGroupName         string                 `json:"quota-group"`
```
Comment (pedronis) [inline daemon/api_snaps.go:206]: validate() should check that this is set only when it makes sense?

File: overlord/servicestate/quota_handlers.go [hunk 3]
```diff
@@ -416,8 +416,8 @@ func quotaRemove(st *state.State, action QuotaControlAction, allGrps map[string]
 		return nil, nil, false, fmt.Errorf("internal error, AddSnaps option cannot be used with remove action")
 	}
 
-	if action.ResourceLimits.Memory != nil {
-		return nil, nil, false, fmt.Errorf("internal error, MemoryLimit option cannot be used with remove action")
+	if !action.ResourceLimits.IsZero() {
+		return nil, nil, false, fmt.Errorf("internal error, quota limit options cannot be used with remove action")
```
Comment (pedronis) [inline overlord/servicestate/quota_handlers.go:420]: I looked at these changes in #12093

File: cmd/snap/cmd_snap_op_test.go [hunk 4]
```diff
@@ -419,6 +419,28 @@ func (s *SnapOpSuite) TestInstallDevMode(c *check.C) {
 	c.Check(s.srv.n, check.Equals, s.srv.total)
 }
 
+func (s *SnapOpSuite) TestInstallQuotaGroup(c *check.C) {
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op_test.go:422]: we also want a TestInstallPath* variant as we have for other options

**Review by pedronis — COMMENTED** [review]
thank you, did a first pass

File: overlord/servicestate/quota_handlers.go (hunk 3 above)
Comment (Meulengracht) [reply overlord/servicestate/quota_handlers.go:420]: Yea you are right, I accidently included that PR, i'll just remove it

File: client/snap_op_test.go [hunk 5]
```diff
@@ -470,6 +470,37 @@ func (cs *clientSuite) TestClientOpInstallPathManyWithOptions(c *check.C) {
...
+		QuotaGroupName: "foo-group",
+	})
+
+	c.Assert(err, check.IsNil)
+
+	body, err := ioutil.ReadAll(cs.req.Body)
+	c.Assert(err, check.IsNil)
+
+	c.Check(string(body), check.Matches, `(?s).*Content-Disposition: form-data; name="dangerous"\r\n\r\ntrue\r\n.*`)
+	c.Check(string(body), check.Matches, `(?s).*Content-Disposition: form-data; name="quota-group"\r\n\r\nfoo-group\r\n.*`)
```
Comment (mardy) [inline client/snap_op_test.go:501]: nitpick: I _think_ you should be able to do
```suggestion
	c.Check(string(body), check.Contains, "Content-Disposition: form-data; name=\"dangerous\"\r\n\r\ntrue\r\n")
	c.Check(string(body), check.Contains, "Content-Disposition: form-data; name=\"quota-group\"\r\n\r\nfoo-group\r\n")
```

File: daemon/api_snaps.go [hunk 6]
```diff
@@ -269,6 +271,9 @@ func (inst *snapInstruction) validate() error {
 	default:
 		return fmt.Errorf("invalid value for transaction type: %s", inst.Transaction)
 	}
+	if inst.QuotaGroupName != "" && inst.Action != "install" {
+		return fmt.Errorf("quota-group can only be specified for install")
```
Comment (mardy) [inline daemon/api_snaps.go:275]: nitpick:
```suggestion
		return fmt.Errorf("quota-group can only be specified on install")
```

File: tests/main/snap-quota-install/task.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
+summary: Test for assigning quota group on snap install.
+
+details: |
+  Test support for assigning a snap quota group on install of a snap
+prepare: |
+  snap pack "$TESTSLIB"/snaps/basic
+  snap set system experimental.quota-groups=true
+  tests.cleanup defer snap unset system experimental.quota-groups
+execute: |
```
Comment (mardy) [inline tests/main/snap-quota-install/task.yaml:9]: we usually have a blank line before yaml keys

File: tests/main/snap-quota-install/task.yaml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  echo "Checking quota group details"
+  snap quota group-one | cat -n > details.txt
+  MATCH "     1\s+name:\s+group-one$" < details.txt
+  MATCH "     2\s+constraints:$" < details.txt
+  MATCH "     3\s+memory:\s+400MB$" < details.txt
+  MATCH "     4\s+current:$" < details.txt
+  MATCH "     5\s+memory:\s+[0-9.a-zA-Z]+B$" < details.txt
+  MATCH "     6\s+snaps:$" < details.txt
+  MATCH "     7\s+-\s+hello-world$" < details.txt
+  echo "Removing the snap from the system again"
```
Comment (mardy) [inline tests/main/snap-quota-install/task.yaml:36]: I guess that "again" is a leftover?

**Review by mardy — APPROVED** [review]
LGTM, just super nitpicks!

File: cmd/snap/cmd_snap_op_test.go (hunk 4 above)
Comment (Meulengracht) [reply cmd/snap/cmd_snap_op_test.go:422]: Fixed! Thank you these notes, greatly appreciated!

File: daemon/api_snaps.go (hunk 2 above)
Comment (Meulengracht) [reply daemon/api_snaps.go:206]: I added one that checks for the option being present only for install command

File: daemon/api_sideload_n_try.go (hunk 1 above)
Comment (Meulengracht) [reply daemon/api_sideload_n_try.go:162]: Done!

File: client/snap_op_test.go (hunk 5 above)
Comment (Meulengracht) [reply client/snap_op_test.go:501]: My go might be completely off, but go is insisting there is no `check.Contains` ?

File: daemon/api_snaps.go (hunk 6 above)
Comment (Meulengracht) [reply daemon/api_snaps.go:275]: Great suggestion, thanks!

File: tests/main/snap-quota-install/task.yaml (hunk 8 above)
Comment (Meulengracht) [reply tests/main/snap-quota-install/task.yaml:36]: Yea thank you for noticing this, it removed my formatting when I copied the file from one branch to the other. Just restored syntax and fixed that

File: client/snap_op_test.go (hunk 5 above)
Comment (pedronis) [reply client/snap_op_test.go:501]: we have a testutil.Contains

**Review by pedronis — APPROVED** [review]
thank you

File: client/snap_op_test.go (hunk 5 above)
Comment (mvo5) [reply client/snap_op_test.go:502]: I like this and update the code

Comment (Meulengracht) [reply client/snap_op_test.go:502]: Ah, thank you both

**Pull request merged**